### PR TITLE
Restore splitter persistence

### LINKED ISSUE #8745: Splitter positions are not restored correctly
## Current behaviour

The splitter positions of FormBrowse are not re/stored correctly. At least the CommitInfo panel is shrinked by 8 pixels if displayed beside the RevisionGrid.

## Expected behaviour

CommitInfo width is restored - all splitter positions of course.

## Steps to reproduce

- position CommitInfo beside RevisionGrid
- restart GE

## Did this work in previous version of GitExtensions

yes, before borders were added

## Environment

- Git Extensions 33.33.33
- Build d2830009cd34a660040c66e49a68e7717cd3b2d4
- Git 2.27.0.windows.1 (recommended: 2.28.0 or later)
- Microsoft Windows NT 10.0.19042.0
- .NET Framework 4.8.4300.0
- DPI 96dpi (no scaling)

### LINKED ISSUE #9000: Toggle left panel status reset after restart
## Current behaviour
Toggle left panel status reset after restart

## Expected behaviour
Toggle left panel status has to be same after restart

## Steps to reproduce
1. turn off left panel
2. restart Git Extensions

## Did this work in previous version of GitExtensions
Yes, but it was long ago. I don't remember the version

## Environment
- Git Extensions 3.6.0.11442
- Build 549f02363b5470696a27a1f002241f5ae4f5c719 (Dirty)
- Git 2.30.2.windows.1
- Microsoft Windows NT 10.0.19042.0
- .NET Framework 4.8.4300.0
- DPI 96dpi (no scaling)

### LINKED ISSUE #9089: Left panel toggle state is not stored/restored in 3.5 RC 1
## Current behaviour

Changing toggle for left panel in main window to disabled is not stored and restored at next start

## Expected behaviour

State of left panel toggle should be restored on next start

## Steps to reproduce

- disable left panel
- restart GitExtensions
-> panel is visible again

Split view layout and Commit info position is stored correct

## Did this work in previous version of GitExtensions

Yes: latest stable release (3.4)

## Environment

- Git Extensions 3.5.0.11509
- Build 009d10c765a90aa00e043e40d5117256b793bc8c (Dirty)
- Git 2.31.1.windows.1
- Microsoft Windows NT 10.0.18363.0
- .NET Framework 4.8.4300.0
- DPI 144dpi (150% scaling)

### PR DESCRIPTION
**Pull request by RussKie**
Resolves #8745
Resolves #9000
Resolves #9089

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/SplitterManager.cs [hunk 1]
```diff
@@ -28,19 +28,19 @@ public void AddSplitter(SplitContainer splitter, string settingName, int? defaul
         public void RestoreSplitters()
         {
             // TODO: Disabled as splitters have become unstable. Refer to #8745 for more details.
```
Comment (mstv) [inline GitUI/SplitterManager.cs:30]: This comment should be deleted, too.

BTW: Never commit/merge code commented out using line comments (`////` in this case) again. Please use `#if false` ... `#endif` in order to get usable diffs.

File: GitUI/SplitterManager.cs [hunk 2]
```diff
@@ -28,19 +28,19 @@ public void AddSplitter(SplitContainer splitter, string settingName, int? defaul
...
-            ////}
+            foreach (var splitter in _splitters)
+            {
+                splitter.RestoreFromSettings(_settings);
+            }
         }
 
         public void SaveSplitters()
         {
             // TODO: Disabled as splitters have become unstable. Refer to #8745 for more details.
```
Comment (mstv) [inline GitUI/SplitterManager.cs:39]: ditto

File: UnitTests/GitUI.Tests/SplitterManagerTest.cs [hunk 3]
```diff
@@ -6,216 +6,215 @@
...
+                splitManager.AddSplitter(splitter, splitterName);
+                int splitterNewWidth = splitterWidth + deltaWidth;
+                splitter.Width = splitterNewWidth;
+                splitter.FixedPanel = FixedPanel.Panel2;
+                splitManager.RestoreSplitters();
+
+                // assert splitter moved by the width delta
+                // Note: if the width of the splitter control itself is not regarded,
+                // this splitter control would move to the left every time the splitter container is restored.
+                splitter.SplitterDistance.Should().Be(splitterDistance + deltaWidth + splitter.SplitterWidth - 2);
```
Comment (mstv) [inline UnitTests/GitUI.Tests/SplitterManagerTest.cs:112]: A comment should explain the `- 2` including an issue number.

File: GitUI/CommandsDialogs/FormCommit.Designer.cs [hunk 4]
```diff
@@ -184,9 +184,9 @@ private void InitializeComponent()
             this.toolbarCommit.SuspendLayout();
             this.commitStatusStrip.SuspendLayout();
             this.SuspendLayout();
-            //
+            //
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormCommit.Designer.cs:187]: Have you added the trailing spaces on purpose?

**Review by mstv — COMMENTED** [review]
Better, but #8745 is not solved.
I have the CommitInfo panel located right of the RevisionGrid. It is shrinked by 4 pixels now on every start.

**Review by ivangrek — COMMENTED** [review]
Commit info panel slides to the right.

File: UnitTests/CommonTestUtils/MemorySettings.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+﻿#nullable enable
+
+using GitCommands.Settings;
+using GitUIPluginInterfaces;
+
+namespace GitUITests
+{
+    public class MemorySettings : SettingsContainer<MemorySettings, MemorySettingsCache>
```
Comment (ivangrek) [inline UnitTests/CommonTestUtils/MemorySettings.cs:8]: Instead, we could use a temporary file or some kind of mock.

**Comment by RussKie** [comment]
Feel free to propose an alternative.

File: GitUI/CommandsDialogs/FormCommit.Designer.cs (hunk 4 above)
Comment (RussKie) [reply GitUI/CommandsDialogs/FormCommit.Designer.cs:187]: No, this is the "designer special" 😒 
/cc: @merriemcgaw

**Comment by RussKie** [comment]
> Commit info panel slides to the right.

Hmm... I haven't observed this in the default layout during local testing (comparing screenshots before and after). What layout are you using?

File: GitUI/SplitterManager.cs (hunk 1 above)
Comment (RussKie) [reply GitUI/SplitterManager.cs:30]: Aye, that's a good point to remember for all of us.
Thanks

**Comment by mstv** [comment]
> What layout are you using?

> I have the CommitInfo panel located right of the RevisionGrid. It is shrinked by 4 pixels now on every start.

I have prepared a similar workaround for 3.5.1.

**Comment by mstv** [comment]
Handle also the following controls, please:
- CommitInfo left / right of the graph, e.g. https://github.com/gitextensions/gitextensions/commit/13caf92691cbefdf05cebb31454d7a8c0c7c3264
- splitter position and collapsed state of BlameControl, which is added by #9424

**Comment by RussKie** [comment]
> * CommitInfo left / right of the graph, e.g. [13caf92](https://github.com/gitextensions/gitextensions/commit/13caf92691cbefdf05cebb31454d7a8c0c7c3264)

The test assert stability of the geometry. I tested manually comparing screenshots before/after and everything appears stable.



> * splitter position and collapsed state of BlameControl, which is added by [RevFileTree: Blame instead of View #9424](https://github.com/gitextensions/gitextensions/pull/9424)

The new functionality is outside the scope of the PR, which I plan to service back to 3.5 branch.
@gerhardol will need to add splitter handling to #9424, if necessary.

**Comment by gerhardol** [comment]
> > * splitter position and collapsed state of BlameControl, which is added by [RevFileTree: Blame instead of View #9424](https://github.com/gitextensions/gitextensions/pull/9424)
> 
> The new functionality is outside the scope of the PR, which I plan to service back to 3.5 branch.
> @gerhardol will need to add splitter handling to #9424, if necessary.

That is not new, Blame control in FileHistory is not persisting the splitter
Will probably be separate from #9424

**Comment by RussKie** [comment]
Just discovered that the fix is partial, it works if the app is launched with a repo. If the app is started with the dashboard we never load the splitter information :(
image

**Pull request merged**